src/dependencies.js

```diff
@@ -84,33 +84,64 @@ const getFileDependencies = async function(path, packageJson, state) {
   const dependencies = precinct.paperwork(path, { includeCore: false })
 
   const depsPaths = await Promise.all(
-    dependencies.map(dependency => getImportDependencies(dependency, basedir, packageJson, state))
+    dependencies.map(dependency =>
+      getImportDependencies({ dependency, basedir, packageJson, state, treeShakingModule })
+    )
   )
   return [].concat(...depsPaths)
 }
 
-// `require()` statements can be either `require('moduleName')` or
-// `require(path)`
-const getImportDependencies = function(dependency, basedir, packageJson, state) {
-  if (LOCAL_IMPORT_REGEXP.test(dependency)) {
-    return getLocalImportDependencies(dependency, basedir, packageJson, state)
+const isTreeShakingModule = function(moduleName, treeShakingModule) {
+  return moduleName === treeShakingModule
+}
+
+const getImportDependencies = function({ dependency, basedir, packageJson, state, treeShakingModule }) {
+  const moduleName = getModuleName(dependency)
+  const isLocalImport = LOCAL_IMPORT_REGEXP.test(dependency)
+  if (isLocalImport || isTreeShakingModule(moduleName, treeShakingModule) || isTreeShakable(dependency)) {
+    const treeShakingState = isLocalImport ? { treeShakingModule } : { treeShakingModule: moduleName }
+    return getTreeShakedDependencies({ dependency, basedir, packageJson, state, ...treeShakingState })
   }
 
-  return getModuleDependencies(dependency, basedir, state, packageJson)
+  return getAllDependencies(dependency, basedir, state, packageJson)
 }
 
 const LOCAL_IMPORT_REGEXP = /^(\.|\/)/
 
+const isTreeShakable = function(dependency) {
+  return TREE_SHAKABLE_DEPENDENCIES.has(dependency)
+}
+
+const TREE_SHAKABLE_DEPENDENCIES = new Set([
+  'next/dist/next-server/lib/constants',
+  'next/dist/next-server/lib/constants.js',
+  'next/dist/next-server/lib/document-context.js',
+  'next/dist/next-server/lib/head.js',
+  'next/dist/next-server/lib/i18n/normalize-locale-path',
+  'next/dist/next-server/lib/router/utils/get-route-from-asset-path',
+  'next/dist/next-server/lib/router/utils/path-match',
+  'next/dist/next-server/lib/router/utils/prepare-destination',
+  'next/dist/next-server/lib/utils',
+  'next/dist/next-server/lib/utils.js',
```

Comment:
Note: we could strip the file extension first, so we don't need to have duplicate entries with/without the file extension.